### DISCUSSION TITLE
Add partial verification to go-tuf?

### DISCUSSION THREAD

**Discussion by hosseinsia** [theupdateframework/go-tuf#187] [Ideas]
[Partial verification](https://uptane.github.io/uptane-standard/uptane-standard.html#partial_verification) is part of Uptane specification. This can be implemented easily in go-tuf. I think go-tuf is the best place to put it. What does this community think about adding the partial verification in this repo?

**Comment by trishankatdatadog** [comment]
I don't see a problem as long as we can conditionally compile it, and I'm sure it'll meet some use cases for the public. Do any of the maintainers have any objections?

**Comment by mnm678** [comment]
I'm curious about what partial verification would mean in the context of TUF. In Uptane, the biggest change is that it does not download data from the Image repository, but only from the Director repository. Partial verification ECUs are also able to skip some checks because they are already done by the primary ECU, and I would want to see some equivalent of that for a TUF implementation to skip any step in the verification process.

**Reply by hosseinsia** [reply]
I would say we can make the partial verification a flexible API (rather than providing a specific definition to what involves in a partial verification). So the caller can order required checks, instead of having a regular monolithic check. For example, a user might desire skipping timestamp/snapshot or both (already desired to ignore both for Director repo in Uptane), or to check root or not. WDYT?

**Reply by mnm678** [reply]
As long as we make it clear to users that partial verification is not a substitute for full verification, and should only be used when paired with a trusted device doing full verification, I don't have any objections.

**Comment by JustinCappos** [comment]
The motivation for partial verification is to support extremely weak
devices that perhaps cannot store ~100KB of information or perform more
than 1-2 public key verifications upon boot.  I don't know that we have
that use case for go-tuf users, right?

**Reply by hosseinsia** [reply]
I agree with you! Even conditional compilation wouldn't completely help satisfy the processing power constrains you mentioned. In the use-case we are dealing with, devices do not have limitations you mentioned. However, we already have an implementation for the partial verification and want to share it with the community.
I believe go-tuf is a good place to share the partial verification capability with the rest of community. WDYT?

**Reply by trishankatdatadog** [reply]
> I don't know that we have that use case for go-tuf users, right?

Consider vehicular use cases (i.e., Uptane) that may wish to use go-tuf.

**Comment by JustinCappos** [comment]
Perhaps this belongs under the name Uptane in some way instead...

**Reply by trishankatdatadog** [reply]
> Perhaps this belongs under the name Uptane in some way instead...

I think just a name like `partial_verification()` should go a long way, along with a comment warning that most users should not use it. And, again, it should be conditionally compiled (i.e., not available by default unless you pass, say, an `--uptane` build flag).
